piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+        assert_eq!(full_layout.hit_test_text_position(2, true).map(|p| p.point.x as f64), Some(pie_width));
+        assert_eq!(full_layout.hit_test_text_position(1, true).map(|p| p.point.x as f64), Some(pi_width));
+        assert_eq!(full_layout.hit_test_text_position(0, true).map(|p| p.point.x as f64), Some(p_width));
+
+        assert_eq!(full_layout.hit_test_text_position(0, false).map(|p| p.point.x as f64), Some(null_width));
+        assert_eq!(full_layout.hit_test_text_position(9, true).map(|p| p.point.x as f64), Some(full_layout.width()));
+        assert_eq!(full_layout.hit_test_text_position(10, true).map(|p| p.point.x as f64), None);
+    }
+
+    #[test]
+    fn test_hit_test_text_position_complex_0() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        let input = "é";
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(layout.width()));
+        assert_eq!(input.len(), 2);
+
+        // unicode segmentation is wrong on this one for now.
+        //let input = "🤦\u{1f3fc}\u{200d}\u{2642}\u{fe0f}";
+
+        //let mut text_layout = D2DText::new();
+        //let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        //let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        //assert_eq!(input.graphemes(true).count(), 1);
+        //assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point_x as f64), Some(layout.width()));
+        //assert_eq!(input.len(), 17);
+
+        let input = "\u{0023}\u{FE0F}\u{20E3}"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        assert_eq!(layout.hit_test_text_position(0, true).map(|p| p.point.x), Some(layout.width()));
+        assert_eq!(input.len(), 7);
+        assert_eq!(input.chars().count(), 3);
+    }
+
+    #[test]
+    fn test_hit_test_text_position_complex_1() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+
+        // Notes on this input:
+        // 5 code points
+        // 5 utf-16 code units
+        // 10 utf-8 code units (2/1/3/3/1)
+        // 3 graphemes
+        let input = "é\u{0023}\u{FE0F}\u{20E3}1"; // #️⃣
+
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+        let layout = text_layout.new_text_layout(&font, input).build().unwrap();
+
+        let test_layout_0 = text_layout.new_text_layout(&font, "é").build().unwrap();
+        let test_layout_1 = text_layout.new_text_layout(&font, "é\u{0023}\u{FE0F}\u{20E3}").build().unwrap();
```

Comment:
Would it be better to slice the source string, rather than use a new literal?

Reply:
Perhaps. This was for my own clarity while reading the tests, but I'm happy to change.